### ISSUE TITLE
Moving from plain array to mongoose.Types.DocumentArray

### ISSUE THREAD

**Issue by Dezzymei** [francescov1/mongoose-tsgen#108]
Hello! I am having issues trying to create a new Document in ts (it works fine in js).

`"mongoose": "^5.11.10",`
`"mongoose-tsgen": "7.1.3",`
`node: v16.15.1`
`npm: v8.11.0`

Schema `user.ts`: 
```
...
tikTokAdvertiserAccounts: [
    {
      id: String,
      name: String,
    },
  ],
...
```

I get this in my `types.gen.ts`:
```
export interface UserTikTokAdvertiserAccount {
  id?: string;
  name?: string;
  _id: mongoose.Types.ObjectId;
}
```

I want to create it in my router:
```
    let advertiserAccounts: UserTikTokAdvertiserAccount[] = [];
    advertiserAccounts.push({id:'testid', name:'testname'});

    user.tikTokAdvertiserAccounts = advertiserAccounts; // problem line

    await user.save();
```

This throws an error:
`Type 'UserTikTokAdvertiserAccount[]' is missing the following properties from type 'DocumentArray<UserTikTokAdvertiserAccountDocument>': isMongooseDocumentArray, create, id, $pop, and 8 more.`

So my question is how do I create this DocumentArray?

Interestingly if I try:
```
    user.tikTokAdvertiserAccounts = new mongoose.Types.DocumentArray(
      advertiserAccounts
    );
```
it seems to work out fine without Typescript complaining.

But if I try something on multiple lines then typescript gets confused:
```
    const explicitAccounts = new mongoose.Types.DocumentArray(
      advertiserAccounts
    );
    user.tikTokAdvertiserAccounts = explicitAccounts;
```

`Type 'DocumentArray<Document<any, any, any>>' is not assignable to type 'DocumentArray<UserTikTokAdvertiserAccountDocument>'.
  Type 'Document<any, any, any>' is missing the following properties from type 'UserTikTokAdvertiserAccountDocument': ownerDocument, parent, parentArray`

So my question is how should I create these sub document arrays on this user document?

**Comment by francescov1** [comment]
Sorry for the delay @Dezzymei, yes arrays can be tricky here with Typescript.

I find the best thing to do is always use `push` directly on the Mongoose array.

Eg

```
user.tikTokAdvertiserAccounts.push ({id:'testid', name:'testname'})

await user.save();
```

I don't have a computer to test this right now, so let me know if you still have issues and I'll get back to you ASAP.

**Comment by francescov1** [comment]
Mongoose arrays get initiated with an empty array by default, so you don't even need to initialize it to an empty array.

**Comment by Dezzymei** [comment]
This seems to indeed work, but how can I initialize it? 
Or how can I make the array empty?

i.e. I would have done:
```
user.tikTokAdvertiserAccounts = [];
```

**Comment by Dezzymei** [comment]
I also have an array of strings in here but I have no idea how to create it from scratch without using something like this:

```
new mongoose.Types.Array(string[])
```

**Comment by francescov1** [comment]
Hmm yeah I see your point. You could use the generated array type to ensure type safety, and then push into the Mongoose array like so:

```
let advertiserAccounts: UserTikTokAdvertiserAccount[] = [];
advertiserAccounts.push({id:'testid', name:'testname'});

user.tikTokAdvertiserAccounts.push(...advertiserAccounts)
```

for setting to an empty array, you may have to do some type casting, or just `as any`. Havent explored that too much but can look into it if you are still having issues.

**Comment by Dezzymei** [comment]
I don't really like that solution, but I am guessing this is perhaps more of a `mongoose` problem than your library.

If you do have a solution I would love to implement it because right now it doesn't feel very typesafe for these properties...

**Comment by francescov1** [comment]
Yes unfortunately this is something I haven't been able to get around with Mongoose. I would look into the [Mongoose docs for Typescript usage](https://mongoosejs.com/docs/typescript.html), its possible that things have changed since I last dove into it.

In terms of type safety, I think the solution above is decently safe; you are still using `UserTikTokAdvertiserAccount` to ensure the items in the array are of the correct type. But its possible to forget to use this type, so definitely not a perfect solution.

**Comment by maxshugar** [comment]
Hey, great library! @francescov1 

Any update on this issue? @Dezzymei

**Comment by maxshugar** [comment]
I have the following model:

`
const venueSchema: Schema = new Schema({
	openingHours: [
		{
			day: {
				type: String,
				required: true,
			},
			active: {
				type: Boolean,
				required: true,
			},
			openingTime: {
				type: String,
				required: true,
			},
			closingTime: {
				type: String,
				required: true,
			},
		},
	],
});
`

Which generates:

`
export type VenueDocument = mongoose.Document<
  mongoose.Types.ObjectId,
  VenueQueries
> &
  VenueMethods & {
    openingHours: mongoose.Types.DocumentArray<VenueOpeningHourDocument>;
    _id: mongoose.Types.ObjectId;
  };
`

When I try and set the value of opening hours, I receive a similar error:

`
const venue = await VenueModel.findById(venueId);
venue.openingHours = [
		{
			day: 'Monday',
			active: true,
			openingTime: "09:00",
			closingTime: "17:00",
		}
	];
`

`
Type '{ day: string; active: true; openingTime: string; closingTime: string; }' is not assignable to type 'VenueOpeningHourDocument'.
  Type '{ day: string; active: true; openingTime: string; closingTime: string; }' is missing the following properties from type 'Subdocument<ObjectId, any, any>': $isSingleNested, ownerDocument, parent, $parent, and 50 more.
`

**Comment by francescov1** [comment]
hey @maxshugar let me revisit this in the next few days, i'll see if there's been any updates to Mongoose typing to provide a better solution here.

**Comment by francescov1** [comment]
Let me know your thoughts @Dezzymei @maxshugar, happy to explore other options if you have any ideas, otherwise will close this out if the solution above works

**Comment by maxshugar** [comment]
Thanks @francescov1, I don't mind not using the short hand form.
